### PR TITLE
Expand 'Setting up Developer Environment' in CONTRIBUTING.md

### PR DESCRIPTION
**Pull request by yuvipanda**
Revamping and expanding our contribution docs, to make life
easier for our Outreachy participants.

Ref #2160

### CODE REVIEW

**Comment by yuvipanda** [comment]
Would love a quick review & merge, since our Outreachy participants are already looking through this.

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -6,23 +6,119 @@ JupyterHub also follows the Jupyter [Community Guides](https://jupyter.readthedo
...
-```bash
-git clone https://github.com/jupyterhub/jupyterhub
-cd jupyterhub
-npm install -g configurable-http-proxy
-pip3 install -r dev-requirements.txt -e .
-```
+JupyterHub can only run on MacOS or Linux operating systems. If you
+are using Windows, we recommend using [VirtualBox](https://www.virtualbox.org/)
+or a similar system to run [Ubuntu Linux](https://www.ubuntu.com/)
+for development.
```
Comment (consideRatio) [inline CONTRIBUTING.md]: EXCELLENT CLARIFICATION!

YES CAPS.

File: CONTRIBUTING.md [hunk 2]
```diff
@@ -6,23 +6,119 @@ JupyterHub also follows the Jupyter [Community Guides](https://jupyter.readthedo
...
+it is to use [miniconda](https://conda.io/miniconda.html). Remember
+to get the 'Python 3' version, and **not** the 'Python 2' version!
+
+### Install nodejs
+
+`configurable-http-proxy`, the default proxy implementation for JupyterHub,
+is written in Javascript to run on [NodeJS](https://nodejs.org/en/). If
+you have not installed nodejs before, we recommend installing it in the
+`miniconda` environment you set up for Python. You can do so with
+`conda install nodejs`.
```
Comment (consideRatio) [inline CONTRIBUTING.md]: Oh, cool, thanks for that advice.

File: CONTRIBUTING.md [hunk 3]
```diff
@@ -6,23 +6,119 @@ JupyterHub also follows the Jupyter [Community Guides](https://jupyter.readthedo
...
+JupyterHub uses [git](https://git-scm.com) & [GitHub](https://github.com)
+for development & collaboration. You need to [install git](https://git-scm.com/book/en/v2/Getting-Started-Installing-Git)
+to work on JupyterHub. We also recommend getting a free account
+on GitHub.com.
+
+## Set up a development install
+
+When developing JupyterHub, you need to make changes to the code
+& see their effects quickly. You need to do a developer install
+to make that happen.
```
Comment (consideRatio) [inline CONTRIBUTING.md]: This is awesome, you write excellent documentation.

File: CONTRIBUTING.md [hunk 4]
```diff
@@ -6,23 +6,119 @@ JupyterHub also follows the Jupyter [Community Guides](https://jupyter.readthedo
...
+   pip3 install -r dev-requirements.txt
+   pip3 install -r requirements.txt
+   ```
+
+5. Install the development version of JupyterHub. This lets you edit
+   JupyterHub code in a text editor & restart the JupyterHub process
+   to see your code changes immediately.
+
+   ```bash
+   pip3 install --editable .
```
Comment (consideRatio) [inline CONTRIBUTING.md]: Thanks for writing out `--editable` rather than `-e`! Especially in this context.

File: CONTRIBUTING.md (hunk 1 above)
Comment (yuvipanda) [reply CONTRIBUTING.md]: :D glad you like it!

File: CONTRIBUTING.md (hunk 3 above)
Comment (yuvipanda) [reply CONTRIBUTING.md]: Thank you, this makes me feel very happy & appreciated <3

**Pull request merged**